packages/apollo-engine-reporting/src/schemaReporter.ts

```diff
@@ -0,0 +1,166 @@
+        reportServerInfo(info: $info, executableSchema: $executableSchema) {
+          inSeconds
+          withExecutableSchema
+        }
+      }
+    }
+  }
+`;
+
+export function reportingLoop(
+  schemaReporter: SchemaReporter,
+  logger: Logger,
+  sendNextWithExecutableSchema: boolean,
+  fallbackReportingDelayInMs: number,
+) {
+  function inner() {
+    // Bail out permanently
+    if (schemaReporter.stopped()) return;
+
+    // Not awaiting this. The callback is handled in the `then` and it calls inner()
+    // to report the server info in however many seconds we were told to wait from
+    // Apollo Graph Manager
+    schemaReporter
+      .reportServerInfo(sendNextWithExecutableSchema)
+      .then(({ inSeconds, withExecutableSchema }) => {
+        sendNextWithExecutableSchema = withExecutableSchema;
+        setTimeout(inner, inSeconds * 1000);
+      })
+      .catch((error: any) => {
+        // In the case of an error we want to continue looping
+        // We can add hardcoded backoff in the future,
+        // or on repeated failures stop responding reporting.
+        logger.error(
+          `Error in reporting server info to Apollo Graph Manager for schema reporting: ${error}`,
+        );
+        sendNextWithExecutableSchema = false;
+        setTimeout(inner, fallbackReportingDelayInMs);
+      });
+  }
+
+  inner();
+}
+
+interface ReportServerInfoReturnVal {
+  inSeconds: number;
+  withExecutableSchema: boolean;
+}
+
+// This class is meant to be a thin shim around the gql mutations.
+export class SchemaReporter {
+  // These mirror the gql variables
+  private readonly serverInfo: EdgeServerInfo;
+  private readonly executableSchemaDocument: any;
+  private readonly url: string;
+
+  private isStopped: boolean;
+  private readonly headers: Headers;
+
+  constructor(
+    serverInfo: EdgeServerInfo,
+    schemaSdl: string,
+    apiKey: string,
+    schemaReportingEndpoint: string | undefined,
+  ) {
+    this.headers = new Headers();
+    this.headers.set('Content-Type', 'application/json');
+    this.headers.set('x-api-key', apiKey);
+    this.headers.set('apollographql-client-name', 'apollo-engine-reporting');
+    this.headers.set(
+      'apollographql-client-version',
+      require('../package.json').version,
+    );
+
+    this.url =
+      schemaReportingEndpoint ||
+      'https://engine-graphql.apollographql.com/api/graphql';
+
+    this.serverInfo = serverInfo;
+    this.executableSchemaDocument = schemaSdl;
+    this.isStopped = false;
+  }
+
+  public stopped(): Boolean {
+    return this.isStopped;
+  }
+
+  public stop() {
+    this.isStopped = true;
+  }
+
+  public async reportServerInfo(
+    withExecutableSchema: boolean,
+  ): Promise<ReportServerInfoReturnVal> {
+    const { data, errors } = await this.graphManagerQuery({
+      info: this.serverInfo,
+      executableSchema: withExecutableSchema
+        ? this.executableSchemaDocument
+        : null,
+    } as ReportServerInfoVariables);
```

Comment:
Is this typecast necessary?  I would somewhat expect it to be inferred by the argument type specified in the signature for `graphManagerQuery` below.  If not, fine to leave it as is, but still surprised because structural typing.

Reply:
Yea it isn't necessary. Removed it.